### DISCUSSION TITLE
Very large transactions and out of memory when using ORM inserts

### DISCUSSION THREAD

**Discussion by electricworry** [sqlalchemy/sqlalchemy#8760] [Usage Questions] [answered]
I realise that this is a bit of a misuse of an ORM, but if I want to `add()` a very large number of rows (e.g. 10 million) in one transaction using a model (i.e. `session.add(some_object)`), then the python process memory usage will grow to the point that the OOM reaper kills the process. This happens even although autoflush is enabled and the data should already be handed to the database.

The same is not true if instead of `add()`, arbitrary queries are run into the session, i.e. `session.execute("INSERT INTO ...")`.

Is there any way to prevent the memory exhaustion in the former case? I'm guessing that the objects cannot be garbage collected because there are still references to them until the transaction is complete. Not even an `expunge_all()` will help, in my testing.

**Comment by zzzeek** [comment, marked as answer]
you probably shouldnt use orm Session.add() to insert 10M rows.   the Session is going to keep track of rows that were added within the transaction as it flushes, so to ensure those get garbage collected youd want to flush every 10K objects or so.

You will get much better performance if you use [bulk_insert_mappings](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.bulk_insert_mappings) with batches of say 10K rows each, and ensure each row has the primary key passed in .  If using 2.0, then you have a lot more options here as performance has been dramatically improved (See https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#optimized-orm-bulk-insert-now-implemented-for-all-backends-other-than-mysql )

**Comment by electricworry** [comment]
Thanks. I agree, I'm doing it in a suboptimal way. :smile: My day job doesn't require me to specialise in databases much so I'm not required to be adept at this. I'll take any advice as better than my default.

So, from what I gather from your first suggestion, autoflush on each `add()` is worse than setting autoflush=False and calling `flush()` on batches? I'll give that a go and see if it has a difference.

I'll also try your `bulk_insert_mappings` suggestion. If I may inquire, how does this work if I'm not including a primary key in the object creation and relying on the primary key (Integer) being auto-generated?

(If neither work to solve the problem, I'll resort to regular INSERTs, as SQL is at least something I understand. :smiley: )

**Reply by zzzeek** [reply]
add() by itself doesnt autoflush.  so if you did add_all(10k objects) then flush(), that might work.  I think the Session should not be holding onto objects after the flush **provided** that you make sure you don't have any references to them in memory on your end.

by all means though using a plain INSERT will be the best performing option short of using the databases bulk copy routines, which are the fastest, if they are available.

**Reply by electricworry** [reply]
Thank you! I did not appreciate that `flush()` is not automatically called for calls to `add()` with autoflush=True. I have confirmed that by introducing a manual `flush()` per 10k objects the memory usage is kept low. My implementation might be suboptimal but this is just something that needs to run once per month.

I'll mark your first response as the answer, and hopefully this discussion might help someone else with this problem.
